### ISSUE TITLE
Triton softmax cause a crash

### ISSUE THREAD

**Issue by nouiz** [google/jax#16973]
### Description

There is a crash related to triton softmax. Disabling it with the `XLA_FLAGS=--xla_gpu_enable_triton_softmax_fusion=false` remove the error.

The error is:
```
======================= test session starts =======================
platform linux -- Python 3.10.6, pytest-7.4.0, pluggy-1.2.0
rootdir: /opt/praxis
plugins: jaxtyping-0.2.20, anyio-3.7.1, typeguard-4.1.0
collected 52 items / 51 deselected / 1 selected

embedding_softmax_test.py Fatal Python error: Aborted

Thread 0x00007fab7d354000 (most recent call first):
  File "/opt/jax-source/jax/_src/interpreters/pxla.py", line 1228 in __call__
  File "/opt/jax-source/jax/_src/profiler.py", line 314 in wrapper
  File "/opt/jax-source/jax/_src/pjit.py", line 1143 in _pjit_call_impl_python
  File "/opt/jax-source/jax/_src/pjit.py", line 1187 in call_impl_cache_miss
  File "/opt/jax-source/jax/_src/pjit.py", line 1203 in _pjit_call_impl
  File "/opt/jax-source/jax/_src/core.py", line 821 in process_primitive
  File "/opt/jax-source/jax/_src/core.py", line 389 in bind_with_trace
  File "/opt/jax-source/jax/_src/core.py", line 2596 in bind
  File "/opt/jax-source/jax/_src/pjit.py", line 165 in _python_pjit_helper
  File "/opt/jax-source/jax/_src/pjit.py", line 252 in cache_miss
  File "/opt/jax-source/jax/_src/traceback_util.py", line 166 in reraise_with_filtered_traceback
  File "/opt/jax-source/jax/_src/numpy/array_methods.py", line 256 in deferring_binary_op
  File "/opt/praxis/praxis/layers/embedding_softmax.py", line 318 in __call__
  File "/opt/flax/flax/linen/module.py", line 967 in _call_wrapped_method
  File "/opt/flax/flax/linen/module.py", line 467 in wrapped_module_method
    File "/opt/flax/flax/linen/module.py", line 2307 in scope_fn
  File "/opt/flax/flax/core/scope.py", line 998 in wrapper
  File "/opt/flax/flax/linen/module.py", line 1682 in apply
  File "/opt/jax-source/jax/_src/traceback_util.py", line 166 in reraise_with_filtered_traceback
  File "/opt/praxis/praxis/base_layer.py", line 1799 in apply
  File "/opt/flax/flax/linen/module.py", line 967 in _call_wrapped_method
  File "/opt/flax/flax/linen/module.py", line 467 in wrapped_module_method
  File "/opt/praxis/praxis/layers/embedding_softmax_test.py", line 374 in test_simple_softmax_layer_class_probs
  File "/usr/local/lib/python3.10/dist-packages/absl/testing/parameterized.py", line 320 in bound_param_test
  File "/usr/lib/python3.10/unittest/case.py", line 549 in _callTestMethod
  File "/usr/lib/python3.10/unittest/case.py", line 591 in run
  File "/usr/lib/python3.10/unittest/case.py", line 650 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/unittest.py", line 333 in runtest
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 169 in pytest_runtest_call
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 80 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 112 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 433 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 262 in <lambda>
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 341 in from_call
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 261 in call_runtest_hook
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 222 in call_and_report
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 133 in runtestprotocol
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 114 in pytest_runtest_protocol
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 80 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 112 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 433 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 349 in pytest_runtestloop
    File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 80 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 112 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 433 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 324 in _main
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 270 in wrap_session
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 317 in pytest_cmdline_main
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 80 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 112 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 433 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/config/__init__.py", line 166 in main
  File "/usr/local/lib/python3.10/dist-packages/_pytest/config/__init__.py", line 189 in console_main
  File "/usr/local/bin/pytest", line 8 in <module>
  
  Extension modules: jaxlib.cpu_feature_guard, numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, zstandard.backend_c, msgpack._cmsgpack, yaml._yaml, scipy._lib._ccallback_c, numpy.linalg.lapack_lite, scipy.sparse._sparsetools, _csparsetools, scipy.sparse._csparsetools, scipy.sparse.linalg._isolve._iterative, scipy.linalg._fblas, scipy.linalg._flapack, scipy.linalg.cython_lapack, scipy.linalg._cythonized_array_utils, scipy.linalg._solve_toeplitz, scipy.linalg._decomp_lu_cython, scipy.linalg._matfuncs_sqrtm_triu, scipy.linalg.cython_blas, scipy.linalg._matfuncs_expm, scipy.linalg._decomp_update, scipy.linalg._flinalg, scipy.sparse.linalg._dsolve._superlu, scipy. ...
s, pandas._libs.window.indexers, pandas._libs.reshape, pandas._libs.groupby, pandas._libs.testing, pandas._libs.parsers, pandas._libs.j
son, sentencepiece._sentencepiece, matplotlib._c_internal_utils, matplotlib._path, kiwisolver._cext, matplotlib._image, scipy.cluster._
vq, scipy.cluster._hierarchy, scipy.cluster._optimal_leaf_ordering (total: 196)
Aborted (core dumped)

```

Repro instruction on a 80G GPU:

```
docker run -it --rm --gpus=all ghcr.io/nvidia/pax:nightly-2023-08-03
cd /opt/praxis/praxis/layers
sed -i 's/(32, 30000)/(65536, 50304)/g' embedding_softmax_test.py # GPT 126M size, b=32, s=2048
sed -i 's/num_classes])/num_classes]).astype(jnp.bfloat16)/g' embedding_softmax_test.py # change to bfloat16
pip install pytest

pytest embedding_softmax_test.py -k test_simple_softmax_layer_class_probs2 # Aborted (core dumped)
XLA_FLAGS="--xla_gpu_enable_triton_softmax_fusion=false" pytest embedding_softmax_test.py -k test_simple_softmax_layer_class_probs2 # Runnable
```

My first guess would be that the grid/block passed isn't the right one.

### What jax/jaxlib version are you using?

upstream

### Which accelerator(s) are you using?

GPU

### Additional system info

Linux

### NVIDIA GPU info

Reproduced on H100 80G. Need 80G to reproduce. Probably reproduce on A100 80G

**Comment by hawkinsp** [comment]
Assigning @sharadmv to follow up, but it's probably @bchetioui who needs to look.

**Comment by bchetioui** [comment]
Managed to reproduce this also on A100-40G by allowing input/output buffers to be aliased.

After investigating and trying to figure out the dimension for which it starts failing, this seems to start failing around shape `(42692, 50304)`. The product of the shape is close to 2**31, so my money is on an integer overflow in Triton itself.

I will follow up there, but our version of Triton in XLA is still behind the [big initial-H100-support commit](https://github.com/openai/triton/commit/f1512bded1934e34f104bf1ac8547e97e24b2fe8), so it might take a little bit of time to get the fix in XLA even if once is fixed upstream, as we need to make sure nothing breaks while integrating H100 support.

**Comment by bchetioui** [comment]
So as it turns out, I gave myself too much credit---the mistake was mine! [This](https://github.com/openxla/xla/pull/4911) should fix it :)

**Comment by nouiz** [comment]
With the container `ghcr.io/nvidia/pax:nightly-2023-08-03`, I get another error now:

```
============================================================== FAILURES ===============================================================
_____________________________________ EmbeddingSoftmaxTest.test_simple_softmax_layer_class_probs2 _____________________________________

self = <praxis.layers.embedding_softmax_test.EmbeddingSoftmaxTest testMethod=test_simple_softmax_layer_class_probs2>
batch_size = 65536, num_classes = 50304

    @parameterized.parameters((8, 1001), (16, 1024), (65536, 50304))
    def test_simple_softmax_layer_class_probs(self, batch_size, num_classes):
    ...
    
      np_get_logits = to_np(logits)
      tf_np_get_logits = to_np(tf_logits)
>     self.assertAllClose(np_get_logits, tf_np_get_logits)

embedding_softmax_test.py:409:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../test_utils.py:87: in assertAllClose
    np.testing.assert_allclose(x, y, rtol=rtol, atol=atol, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<function assert_allclose.<locals>.compare at 0x7f1783bdf910>, array([[ 0.00582462,  0.0245036 ,  0.07030115, ..., -0...],
       [ 0.03172682, -0.02000502,  0.03393763, ..., -0.06318258,
         0.14392377,  0.08553023]], dtype=float32))
kwds = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-05, atol=1e-05', 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError:
E           Not equal to tolerance rtol=1e-05, atol=1e-05
E
E           Mismatched elements: 2345742940 / 3296722944 (71.2%)
E           Max absolute difference: 0.00022475
E           Max relative difference: 179561.94
E            x: array([[ 0.005825,  0.024504,  0.070301, ..., -0.021309,  0.100876,
E                   -0.00767 ],
E                  [-0.083098,  0.137958,  0.125972, ...,  0.10841 ,  0.077706,...
E            y: array([[ 0.005845,  0.024529,  0.070316, ..., -0.02129 ,  0.100827,
E                   -0.007618],
E                  [-0.083073,  0.137941,  0.125988, ...,  0.108402,  0.077672,...

/usr/lib/python3.10/contextlib.py:79: AssertionError
```

**Comment by bchetioui** [comment]
The numbers are not too far off, and this reproduces also with smaller shape (e.g. `(500, 50304)`), with or without Triton Softmax fusion.

This seems like an issue of numerical stability because you changed the test to use `bfloat16` as the initial data type---and thus probably not a bug.

**Comment by nouiz** [comment]
Good catch. Thanks. Closing.
